### PR TITLE
Add basic support for a TeamCity reporter

### PR DESCRIPTION
**Pull request by clhynfield**
This produces pretty rough output and is only available if you supply
the -v option, but in the absence of any documentation on Ginkgo's
TeamCity reporter, it's a start!

### CODE REVIEW

**Comment by clhynfield** [comment]
Doesn't look like this is getting much traction, and in light of the recent [retraction of advice to continuously monitor using these tests](https://github.com/cloudfoundry/cf-smoke-tests/commit/11b6b8f208998dabfbc9534cd49e582c563b8ad1), perhaps it's best to develop separate service level indicator-type tests...

**Comment by dsabeti** [comment]
Hey @clhynfield. We actually had this story prioritized, we just haven't had the bandwidth to get around to merging it. Do you want to re-open the PR? We'll try to get to it soon.

**Comment by clhynfield** [comment]
Oh, cool, @dsabeti – I must have missed that. Yep, I can re-open it, though I'm not sure how much more I should invest in extending this tool given the change in advice for using it. Care to comment on whether this should be used outside something like a post-deploy errand?

**Comment by dsabeti** [comment]
@clhynfield No, it's my bad. There "scheduled" tag is too subtle a way to say that the story was prioritized.

With regards to the smoke-tests being used as a monitoring tool, I think it's a bad idea. Smoke tests perform a `cf push` with a [hard-coded timeout](https://github.com/cloudfoundry/cf-smoke-tests/blob/master/smoke/runtime/init_test.go#L32) and says whether or not the push succeeded. That's really flaky. So the result is binary data with a lot of noise in it -- not a good candidate for monitoring. I'm open to other use cases for smoke-tests, but I think the best use case is for validating the success of the deployment, and monitoring is definitely not a good one.

Did you need this change because you wanted to use smoke-tests as a monitoring tool?

**Comment by jim80net** [comment]
Because some teams are moving away from it, doesn't necessarily make the pattern invalid. 

What some folks have noticed is that smoke tests fail for a large number of reasons, many times with the setup or tear down of the test. This is because there is a testing sensitivity that the platform doesn't actually promise. 

At test is: 
- create space/org
- push app --no-start
- start app
- scale app
- read logs
- stop app
- delete app
- delete space/org

When any one of these steps fails or times out, the whole thing fails. The delete steps are considered asynchronous jobs. We've seen cases where a deployment with a DB migration enqueues jobs in the CC workers, and smoke tests fail as a result of that. Everything was working as designed, but the smoke tests failed. 

There is a mismatch between expected and tested behavior in that example. 

In the referenced commit, @clhynfield mentioned [headandshoulders](https://github.com/onsi/headandshoulders), which could be used to reduce flakiness of setup and tear down.


Further, the instrument triggers false for any number of reasons but only issues a single data point. Certainly in the case of reading logs, the components at test are distinct from what is being tested for scaling (CC -> brain -> BBS -> diego), starting (CC -> blobstore -> buildpacks -> brain -> BBS -> diego ), deleting ( CC -> workers ), etc.. There is room to expand the output of smoke tests to encompass the various facets. 

Pivotal CloudOps is trying out smaller tests. What we're hoping to achieve is something that could deliver SLI (Service Level Indicator) more accurately. Here's what we have so far: [cf-sli](https://github.com/pivotal-cloudops/cf-sli). 

The reason [cf-sli](https://github.com/pivotal-cloudops/cf-sli) works for us, is that we have a project boundary at the setup and tear down steps, so we aren't measuring that. Theoretically, cf-smoke-tests could also do setup and teardown outside of the portion of code that returns a success/fail signal.

**Comment by jim80net** [comment]
TL;DR: 
Test is broad and aggressive with timeouts especially with delete space, producing false negatives. 
Setup and Teardown steps could be separated out and not tested, or produce signals distinct from the core test(s).

**Comment by clhynfield** [comment]
Thanks very much for the elaboration, @jim80net, and I largely agree. I recently switched from an established CF platform team with an extensive observability toolset to a team that's looking for some basic observability into our platforms, and cf-smoke-tests seemed like a quick win. We've had it running hourly against several CF instances for a month or so, and the team is really pleased with the results. A couple of times, it has alerted us to legitimate platform issues, and a couple of times, it's shown there was a problem with our CI platform. 

I've been watching [cf-sli][] since shortly after y'all started it, and it seems like a very narrow slice of a proper SLO-measuring tool, rather than a smoke test. Are you planning to make it extensible, allowing the addition of measurement of other platform features like autoscaling, a full suite of buildpacks, etc? If that's the cf-sli product vision, I'd enthusiastically contribute however I can. 

[cf-sli]: https://github.com/pivotal-cloudops/cf-sli

**Comment by jim80net** [comment]
@njbennett comments about cf-sli vision?

**Comment by njbennett** [comment]
1) We no longer actively recommend running smoketests continuously in production. If you get value from doing so, we don't actively recommend against. 

2) Re: cf-sli vision

This is very much in limbo right now, especially as regards this particular binary. There's a Pivotal team working on a general SLI/SLO dashboard for PCF, and we're reluctant to invest CloudOps capacity in duplicating that work. However, I'm not sure yet how much of that work we're going to be able to use, and when, and whether it covers everything that we need.

In the short term, CloudOps has a few items we want to add to the cf-sli tool and scripts around it (java buildpack, better error output), but otherwise we're focusing on transitioning PWS to `cf-deployment`.

In the mid-to-long term, I definitely have a product vision for some kind of standard set of availability and performance monitors that can be deployed against any cf-deployment-based CF deployment. CloudOps has a need for this thing to improve our management of PWS, but we'd also like to enable better continuous -ility testing in CF component teams, much like cf-deployment and its related tools have enabled more continuous integration testing by those component teams.

I'm not sure if that's the vision for `cf-sli` in particular, and it may in fact end up being provided by that aforementioned Pivotal team. There will hopefully be some more info on that once we've finished the cf-deployment roll-out, and `cf-sli` will continue to be the repository to watch for updates in this area.

**Comment by dsabeti** [comment]
I think @njbennett gave the more nuanced (and accurate) description of my attitude toward smoke-tests:
> We no longer actively recommend running smoketests continuously in production. If you get value from doing so, we don't actively recommend against.

Moreover, until a CFF team (almost certainly the Release Integration team) provides an alternative test suite that we _do_ actively recommend, it seems like using smoke-tests is a sufficient stop-gap measure. Just know that they can be flaky.

As far as this PR goes, I think we should re-open it and we'll get to merging it soon.

**Comment by clhynfield** [comment]
Thanks, team, for the enlightening discussion. @njbennett, it sounds like your platform team is like the ones I've worked with, where I say on the regular, "We'd love to do a better job of this, but we're currently focused on a different MVP."
